add generic x86_32 support

goertzenator opened the pull request:
###### Motivation for this change

Add support for i386-i586. Ref:
- https://discourse.nixos.org/t/i586-cross-compiling-ambiguous-i686-use/1482
- RFC (TBA)

In summary:
- Add `isx86_32` predicate that can replace most uses of `isi686`.
- `isi686` is reinterpreted to mean "exactly i686 arch, and not say i585 or i386".
- This branch was used to build working i586 kernel running on i586 hardware.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

goertzenator committed:
add generic x86_32 support

- Add support for i386-i586.
- Add `isx86_32` predicate that can replace most uses of `isi686`.
- `isi686` is reinterpreted to mean "exactly i686 arch, and not say i585 or i386".
- This branch was used to build working i586 kernel running on i586 hardware.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/i586-cross-compiling-ambiguous-i686-use/1482/8

matthewbauer commented on lib/systems/doubles.nix:
```diff
@@ -12,11 +12,14 @@ let
     "i686-cygwin" "i686-freebsd" "i686-linux" "i686-netbsd" "i686-openbsd"
+    "i586-cygwin" "i586-freebsd" "i586-linux" "i586-netbsd" "i586-openbsd"
+    "i486-cygwin" "i486-freebsd" "i486-linux" "i486-netbsd" "i486-openbsd"
+    "i386-cygwin" "i386-freebsd" "i386-linux" "i386-netbsd" "i386-openbsd"
```
i386 doesn't work with latest glibc - better leave it off

goertzenator replied:
Should I purge all i386 support from the PR or just this i386 line?

matthewbauer replied:
Just the i386-linux references, I think. Although I'm not sure whether other OSs actually still support i386. Very few computers still even exist that are i386 so it's largely a irrelevant 😄

goertzenator replied:
I've followed code to see what `doubles.nix` actually does, and... it does almost nothing.  Only `mesaPlatforms` appears to be used AFAICT.  Anyway I will revert all changes to this file and also `for-meta.nix` (which looks 100% dead).

matthewbauer commented on lib/systems/inspect.nix:
```diff
@@ -8,8 +8,12 @@ let abis = lib.mapAttrs (_: abi: builtins.removeAttrs abi [ "assertions" ]) abis
 
 rec {
   patterns = rec {
+    isi386         = { cpu = cpuTypes.i386; };
```
Do you expect needing these? My thinking would be we would just use x86_32 as much as possible. We are doing a similar thing with isAarch32 as well. Even without these you can always do:

```
stdenv.hostPlatform.parsed.cpu.name == "i386"
```

if you really need a condition on a certain micro architecture.

goertzenator replied:
I don't.  I put them in to be symmetric with `isi686`.

goertzenator committed:
revert `isi[345]86`, remove dead code

- Remove changes to dead code in `doubles.nix` and `for-meta.nix`.
- Remove `isi[345]86` predicates since other cpu families don't have specific model predicates.

goertzenator committed:
remove i386-linux since linux not supported on that cpu

Ericson2314 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -122,7 +122,9 @@ let
-        isi686 isx86_64 is64bit isAarch32 isAarch64 isMips isBigEndian;
+        isi386 isi486 isi586 isi686 isx86_32 isx86_64
+        is32bit is64bit
+        isAarch32 isAarch64 isMips isBigEndian;
```
Also sorry I wrote this (probably on my phone) and then never submitted it.

matthewbauer replied:
Feel free to remove them. At some point we should probably just start removing any stdenv.is* that is unused.
